[java] TestClassWithoutTestCases - can not set test pattern to empty

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

**Affects PMD Version:** 

6.53.0

**Description:**

Docs claim "Test class name pattern to identify test classes by their fully qualified name. An empty pattern disables test class detection by name. Since PMD 6.51.0."

However, adding this to my custom pmd config:

```xml
<rule ref="category/java/errorprone.xml/TestClassWithoutTestCases">
        <properties>
            <property name="testClassPattern" value="" />
        </properties>
    </rule>
```

leads to

**Exception Stacktrace:**

```text
Caused by: java.lang.NullPointerException: Cannot invoke "String.isEmpty()" because "this.pattern" is null
    at java.util.regex.Pattern.<init> (Pattern.java:1428)
    at java.util.regex.Pattern.compile (Pattern.java:1069)
    at net.sourceforge.pmd.properties.ValueParserConstants$10.valueOf (ValueParserConstants.java:211)
    at net.sourceforge.pmd.properties.ValueParserConstants$10.valueOf (ValueParserConstants.java:208)
    at net.sourceforge.pmd.properties.GenericPropertyDescriptor.createFrom (GenericPropertyDescriptor.java:68)
    at net.sourceforge.pmd.properties.AbstractSingleValueProperty.valueFrom (AbstractSingleValueProperty.java:129)
    at net.sourceforge.pmd.rules.RuleFactory.setRulePropertyCapture (RuleFactory.java:307)
    at net.sourceforge.pmd.rules.RuleFactory.setPropertyValues (RuleFactory.java:302)
    at net.sourceforge.pmd.rules.RuleFactory.decorateRule (RuleFactory.java:126)
    at net.sourceforge.pmd.RuleSetFactory.parseRuleReferenceNode (RuleSetFactory.java:726)
    at net.sourceforge.pmd.RuleSetFactory.parseRuleNode (RuleSetFactory.java:505)
    at net.sourceforge.pmd.RuleSetFactory.parseRuleSetNode (RuleSetFactory.java:409)
    at net.sourceforge.pmd.RuleSetFactory.createRuleSet (RuleSetFactory.java:234)
    at net.sourceforge.pmd.RuleSetFactory.createRuleSet (RuleSetFactory.java:229)
    at net.sourceforge.pmd.RuleSetLoader.loadFromResource (RuleSetLoader.java:253)
    at net.sourceforge.pmd.RuleSetLoader.loadFromResource (RuleSetLoader.java:153)
    at net.sourceforge.pmd.RuleSetLoader.loadFromResources (RuleSetLoader.java:185)
    at org.apache.maven.plugins.pmd.exec.PmdExecutor.processFilesWithPMD (PmdExecutor.java:351)
    at org.apache.maven.plugins.pmd.exec.PmdExecutor.run (PmdExecutor.java:248)
    at org.apache.maven.plugins.pmd.exec.PmdExecutor.execute (PmdExecutor.java:83)
    at org.apache.maven.plugins.pmd.PmdReport.executePmd (PmdReport.java:390)
    at org.apache.maven.plugins.pmd.PmdReport.canGenerateReport (PmdReport.java:315)
```

**Code Sample demonstrating the issue:**

not needed.

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  add example above to custom ruleset.
2.  run pmd with the ruleset
3. profit

**Running PMD through:** *[Maven]*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/pmd-core/src/main/java/net/sourceforge/pmd/rules/RuleFactory.java
+++ b/pmd-core/src/main/java/net/sourceforge/pmd/rules/RuleFactory.java
@@ -357,10 +357,8 @@ public class RuleFactory {
 
     /** Gets the string value from a property node. */
     private static String valueFrom(Element propertyNode) {
-        String strValue = propertyNode.getAttribute(DEFAULT_VALUE.attributeName());
-
-        if (StringUtils.isNotBlank(strValue)) {
-            return strValue;
+        if (propertyNode.hasAttribute(DEFAULT_VALUE.attributeName())) {
+            return propertyNode.getAttribute(DEFAULT_VALUE.attributeName());
         }
 
         final NodeList nodeList = propertyNode.getChildNodes();

--- a/pmd-core/src/test/java/net/sourceforge/pmd/RuleSetFactoryTest.java
+++ b/pmd-core/src/test/java/net/sourceforge/pmd/RuleSetFactoryTest.java
@@ -185,7 +185,7 @@ public class RuleSetFactoryTest {
     public void testProps() throws RuleSetNotFoundException {
         Rule r = loadFirstRule(PROPERTIES);
         assertEquals("bar", r.getProperty((PropertyDescriptor<String>) r.getPropertyDescriptor("fooString")));
-        assertEquals(new Integer(3), r.getProperty((PropertyDescriptor<Integer>) r.getPropertyDescriptor("fooInt")));
+        assertEquals(Integer.valueOf(3), r.getProperty((PropertyDescriptor<Integer>) r.getPropertyDescriptor("fooInt")));
         assertTrue(r.getProperty((PropertyDescriptor<Boolean>) r.getPropertyDescriptor("fooBoolean")));
         assertEquals(3.0d, r.getProperty((PropertyDescriptor<Double>) r.getPropertyDescriptor("fooDouble")), 0.05);
         assertNull(r.getPropertyDescriptor("BuggleFish"));
@@ -220,6 +220,28 @@ public class RuleSetFactoryTest {
         PropertyDescriptor<List<String>> prop = (PropertyDescriptor<List<String>>) r.getPropertyDescriptor("packageRegEx");
         List<String> values = r.getProperty(prop);
         assertEquals(Arrays.asList("com.aptsssss", "com.abc"), values);
+    }
+
+    /**
+     * Verifies that empty values for properties are possible. Empty values can be used to disable a property.
+     * However, the semantic depends on the concrete rule implementation.
+     *
+     * @see <a href="https://github.com/pmd/pmd/issues/4279">[java] TestClassWithoutTestCases - can not set test pattern to empty #4279</a>
+     */
+    @Test
+    public void testEmptyStringProperty() throws RuleSetNotFoundException {
+        Rule r = loadFirstRule("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+                + "<ruleset name=\"test\">\n "
+                + " <description>ruleset desc</description>\n     "
+                + "<rule name=\"myRule\" message=\"Do not place to this package. Move to \n{0} package/s"
+                + " instead.\" \n" + "class=\"net.sourceforge.pmd.RuleWithProperties\" language=\"dummy\">\n"
+                + "         <description>Please move your class to the right folder(rest \nfolder)</description>\n"
+                + "         <priority>2</priority>\n         <properties>\n             <property name=\"stringProperty\""
+                + " value=\"\" />\n"
+                + "         </properties></rule>" + "</ruleset>");
+        PropertyDescriptor<String> prop = (PropertyDescriptor<String>) r.getPropertyDescriptor("stringProperty");
+        String value = r.getProperty(prop);
+        assertEquals("", value);
     }
 
     @Test

